### PR TITLE
Document `cml runner` Bitbucket support

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
**Part of https://github.com/iterative/cml/pull/798#pullrequestreview-982578801**
  * https://github.com/iterative/cml/pull/798#issuecomment-1135299662
  * https://github.com/iterative/cml/pull/798#issuecomment-1135335406
  * https://github.com/iterative/cml/pull/798#issuecomment-1135391470

> **Warning**
> The GPU example doesn't work, because of... reasons; see [iterative/cml#1015](https://github.com/iterative/cml/issues/1015)

### CODE REVIEW

**Comment by 0x2b3bfa0** [comment]
CircleCI checks are broken; asked [here](https://iterativeai.slack.com/archives/C01R57GNPNF/p1653391399496199)

**Comment by julieg18** [comment]
> CircleCI checks are broken; asked [here](https://iterativeai.slack.com/archives/C01R57GNPNF/p1653391399496199)

Will be fixed once #244 is merged!

**Review by dacbd — APPROVED** [review]
Simple and reads fine to me.

File: content/docs/self-hosted-runners.md [hunk 1]
```diff
@@ -71,7 +68,7 @@ jobs:
       - uses: actions/checkout@v2
       - name: Train model
         env:
-          REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
```
[the remark this answers is not in the document]
Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md:74]: There is a reason for this change: the `PERSONAL_ACCESS_TOKEN` name is arbitrary, and `cml` only reads the `REPO_TOKEN` environment variable; forges other than GitHub don't have explicit secrets, so it's better to unify the names.

File: content/docs/self-hosted-runners.md [hunk 2]
```diff
@@ -1,18 +1,15 @@
 # Self-hosted (On-premise or Cloud) Runners
 
-GitHub Actions and GitLab CI workflows are run on GitHub- and GitLab- hosted
-runners by default. However, there are many great reasons to use your own
-runners: to take advantage of GPUs, orchestrate your team's shared computing
-resources, or train in the cloud.
+GitHub Actions, GitLab CI/CD and Bitbucket Pipelines workflows are run on
+GitHub-, GitLab- and Bitbucket- hosted runners by default. However, there are
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:4]:
```suggestion
GitHub Actions, GitLab CI/CD, and Bitbucket Pipelines workflows are executed on
"native" runners (hosted by GitHub/GitLab/Bitbucket respectively) by default.
However, there are
```

File: content/docs/self-hosted-runners.md [hunk 3]
```diff
@@ -1,18 +1,15 @@
...
+many great reasons to use your own runners: to take advantage of GPUs,
+orchestrate your team's shared computing resources, or train in the cloud.
 
 ☝️ **Tip!** Check out the official documentation from
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:8]:
```suggestion
<admon type="tip">

Check out the official documentation from
```

File: content/docs/self-hosted-runners.md [hunk 4]
```diff
@@ -1,18 +1,15 @@
...
-and [GitLab](https://docs.gitlab.com/runner) for more information on self-hosted
-runners.
-
-⚠️ Using
-[self-hosted runners for Bitbucket Pipelines](https://support.atlassian.com/bitbucket-cloud/docs/runners)
-is not yet supported.
+[GitHub](https://help.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners),
+[GitLab](https://docs.gitlab.com/runner) and
+[Bitbucket](https://support.atlassian.com/bitbucket-cloud/docs/runners) for more
+information on self-hosted runners.
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:12]:
```suggestion
information on self-hosted runners.

</admon>
```

File: content/docs/self-hosted-runners.md [hunk 5]
```diff
@@ -104,17 +101,40 @@ train-model:
...
+        script:
+          - |
+            cml runner \
+                --cloud=aws \
+                --cloud-region=us-west \
+                --cloud-type=p2.xlarge \
+                --cloud-spot \
+                --labels=cml-gpu
+    - step:
+        runs-on: [self.hosted, cml-gpu]
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:127]: when in Rome?

```suggestion
                --labels=cml.gpu
    - step:
        runs-on: [self.hosted, cml.gpu]
```

File: content/docs/self-hosted-runners.md [hunk 6]
```diff
@@ -104,17 +101,40 @@ train-model:
...
+        image: iterativeai/cml:0-dvc2-base1-gpu
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:128]: do GPUs actually work on BB?

File: content/docs/self-hosted-runners.md [hunk 7]
```diff
@@ -178,16 +198,16 @@ The `cml runner` command supports many options (see the
...
+security), or Bitbucket
+[secured user-defined variables](https://support.atlassian.com/bitbucket-cloud/docs/variables-and-secrets/#User-defined-variables).
 
-⚠️ You will need to create a
-[personal access token (PAT)](#personal-access-token) with repository read/write
-access. In the example workflow above, this token is stored as
-`PERSONAL_ACCESS_TOKEN`.
+⚠️ You will need to create a [personal access token](#personal-access-token)
+with enough permissions to register self-hosted runners. In the example workflow
+above, this token is stored as `REPO_TOKEN`.
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:210]:
```suggestion
<admon type="warn">

You will need to create a [personal access token (PAT)](#personal-access-token)
with enough permissions to register self-hosted runners. In the example workflow
above, this token is stored as `REPO_TOKEN`.

</admon>
```

**Review by casperdcl — CHANGES_REQUESTED** [review]
still todo in follow-ups (https://github.com/iterative/cml/issues/1015?)

- banner announcement
- examples
  + `/start/bitbucket`
  + repos à la https://github.com/iterative/example-repos-dev/issues/100

File: content/docs/self-hosted-runners.md (hunk 6 above)
Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md:133]: No, please see the callout on https://github.com/iterative/cml.dev/pull/243#issue-1246387011

Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md:133]: Should we add a warning and/or remove mentions to GPU until https://github.com/iterative/cml/issues/1015 is resolved?

Comment (casperdcl) [reply content/docs/self-hosted-runners.md:133]: I'd say:

```diff
- --labels=cml.gpu
- --cloud-type=p2.xlarge
- image: iterativeai/cml:0-dvc2-base1-gpu
+ # gpu not yet supported, see https://github.com/iterative/cml/issues/1015
```

File: content/docs/self-hosted-runners.md [hunk 8]
```diff
@@ -104,17 +106,40 @@ train-model:
...
+          - |
+            cml runner \
+                --cloud=aws \
+                --cloud-region=us-west \
+                --cloud-type=p2.xlarge \
+                --cloud-spot \
+                --labels=cml.gpu
+    - step:
+        runs-on: [self.hosted, cml.gpu]
+        image: iterativeai/cml:0-dvc2-base1-gpu
```
Comment (0x2b3bfa0) [inline content/docs/self-hosted-runners.md:133]:
```suggestion
                --cloud-spot
    - step:
        runs-on: [self.hosted]
        image: iterativeai/cml:0-dvc2-base1
```

File: content/docs/self-hosted-runners.md (hunk 6 above)
Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md]: https://github.com/iterative/cml.dev/pull/243#discussion_r882186255

File: content/docs/self-hosted-runners.md [hunk 9]
```diff
@@ -104,17 +106,39 @@ train-model:
...
+```yaml
+pipelines:
+  default:
+    - step:
+        image: iterativeai/cml:0-dvc2-base1
+        script:
+          - |
+            cml runner \
+                --cloud=aws \
+                --cloud-region=us-west \
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:127]:
```suggestion
                --cloud-region=us-west \
                --cloud-type=m5.2xlarge \
```

File: content/docs/self-hosted-runners.md [hunk 10]
```diff
@@ -104,17 +106,39 @@ train-model:
...
+                --cloud-spot
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:128]:
```suggestion
                --cloud-spot \
                --labels=cml
```

File: content/docs/self-hosted-runners.md [hunk 11]
```diff
@@ -104,17 +106,39 @@ train-model:
...
+    - step:
+        runs-on: [self.hosted]
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:130]:
```suggestion
        runs-on: [self.hosted, cml]
```

**Review by casperdcl — CHANGES_REQUESTED** [review]

**Review by casperdcl — APPROVED** [review]
:shipit:

**Pull request merged**